### ISSUE TITLE
Regression: generated Python functions return a list instead of a single value

### ISSUE THREAD

**Issue by hvbtup** [swig/swig#3084]
I use swig 4.3.0 to generate Python wrappers for a C library.

For example, one of the C functions is defined as

    EXPORT PTError API PBMGetLevel(BookmarkHandle h, int* level);

PTError is defined as

    typedef enum PTError {
      PTSuccess = 0,
      PTNotPDF = 1001, ...
    } PTError;

And I am using the following in the *.i file:

    void create_pdf_exception(const char* api, PTError rc);

    %typemap(in,numinputs=0,noblock=1) PTError* err (PTError temp) {
        temp = 0;
        $1 = &temp;
    }

    %typemap(in,numinputs=0,noblock=1) PTError* pErr (PTError temp) {
        temp = 0;
        $1 = &temp;
    }

    %typemap(out) PTError {
        if ($1 == 0 || $1 == PTAlreadyWritten) {
            $result = SWIG_Py_Void();
        } else {
            create_pdf_exception("$symname", $1);
            goto fail;
        }
    }

    %typemap(argout) (PTError* err) {
        if (*$1 == 0 || *$1 == PTAlreadyWritten) {
            ;
        } else {
            Py_XDECREF(resultobj);
            create_pdf_exception("$symname", *$1);
            goto fail;
        }
    }

    %typemap(argout) (PTError* pErr) {
        if (*$1 == 0 || *$1 == PTAlreadyWritten) {
            ;
        } else {
            Py_XDECREF(resultobj);
            create_pdf_exception("$symname", *$1);
            goto fail;
        }
    }

With SWIG 4.1.0, calling the function `PBMGetLevel(bmh)` in Python returned a single integer `-1`.
With SWIG 4.3.0, this now returns a list with two elements: `[None, -1]`.

The same happens for all functions in the C lib:
Instead of just the expected, they return a list containing None and the expected value.

**I think that this issue is a consequence of the fix for #2905.**

Comparing the C code generated with SWIG 4.1.0 and SWIG 4.3.0, it looks like this:

    if (SWIG_IsTmpObj(res2)) {
      resultobj = SWIG_Python_AppendOutput(resultobj, SWIG_From_int((*arg2)), 0);
    } else {
      int new_flags = SWIG_IsNewObj(res2) ? (SWIG_POINTER_OWN |  0 ) :  0 ;
      resultobj = SWIG_Python_AppendOutput(resultobj, SWIG_NewPointerObj((void*)(arg2), SWIGTYPE_p_int, new_flags), 0);
    }

**The additional argument 0 is given to the SWIG_Python_AppendOutput function.**

Is this a bug in SWIG?
If not, what can I do in the *.i file to get the same behavior as in 4.1.0?

**Comment by vadz** [comment]
It looks like the detection of whether the function is `void` or not doesn't work in this case, and I'm not sure if this can be fixed automatically. Being able to mark the function as being `isvoid` would definitely help here...

BTW, do `argout` typemaps actually do anything in this case (and why are there 2 of them?)?

**Comment by hvbtup** [comment]
> BTW, do argout typemaps actually do anything in this case ...

There are other functions in the C lib which use a pointer to a PTError struct as an argument.

> ...(and why are there 2 of them?)?

This is obviously a copy/paste error. It's a duplicate.

**Comment by hvbtup** [comment]
Hmm. 
So it seems there is nothing I can do to make the C lib work with SWIG including error handling?

In this case, I'd have to move the the error handling logic into a higher level Python API which looks at the returned list, pop the first element from it and raise an exception if it's not `None`, with the exception details extracted from the PTError struct.

Or remove SWIG from the build completely, and find a different solution using `ctypes` and `ctypesgen`.

**Comment by vadz** [comment]
> In this case, I'd have to move the the error handling logic into a higher level Python API which looks at the returned list, pop the first element from it and raise an exception if it's not `None`, with the exception details extracted from the PTError struct.

FWIW you should be do this at SWIG level with `%pythoncode`.

**Comment by hvbtup** [comment]
@vadz Thanks, so at least the workaround code can be saved where it belongs.

**Comment by hvbtup** [comment]
Just for information, my workaround code looks like this:


```

%pythoncode %{
"""
The following code is used as a workaround für SWIG issue #3084:
Regression: generated Python functions return a list instead of a single value.

Instead of at the C level, we take care of the PTError return value
at the Python level by using a wrapper function.

"""

def _wrap_error_return(func):
    def wrapped(*args, **kwargs):
        res = func(*args, **kwargs)
        # print(f"{func}({args}, {kwargs}) {res=}")
        if isinstance(res, list):
            if len(res) == 2:
                [err, res] = res
                if err:
                    raise RuntimeError(f"PDF-Error rc={err}")
        elif isinstance(res, int):
                err = res
                if err:
                    raise RuntimeError(f"PDF-Error rc={err}")
                res = None
        # print(f"Returning {res}")
        return res
    return wrapped
    
AddFileAttachment = _wrap_error_return(AddFileAttachment)
AddFileAttachmentW = _wrap_error_return(AddFileAttachmentW)
... many lines like these
%}

```

**Comment by jun66j5** [comment]
I've encountered the same issue while building Python bindings of Subversion 1.14.4 with SWIG 4.3.0.

Example:
```
svn_error_t *
svn_fs_commit_txn(const char **conflict_p, svn_revnum_t *new_rev,
                  svn_fs_txn_t *txn, apr_pool_t *pool)
```
with the following typemaps:
```
%typemap(out) svn_error_t * { ... }
%typemap(argout) const char ** { ... }
%typemap(argout) svn_revnum_t * { ... }
```

As a workaround, the following changed:

- Always initializes the `$result` with an new list in `%typemap(out)`.
- Adjust the `$result` by the count of the list in `%typemap(ret)`.
- Ruby bindings have the same issue, so that the same changes are applied.

```
%typemap(out) svn_error_t * {
    if ($1 != NULL) {
        if ($1->apr_err != SVN_ERR_SWIG_PY_EXCEPTION_SET)
            svn_swig_py_svn_exception($1);
        else
            svn_error_clear($1);
        SWIG_fail;
    }
    Py_XDECREF($result);
    $result = PyList_New(0);
}

%typemap(ret) svn_error_t * {
    if ($result == NULL) {
      $result = Py_None;
      Py_INCREF($result);
    }
    else {
      switch (PyList_Size($result)) {
        case 0:
          $result = Py_None;
          Py_INCREF($result);
          break;
        case 1:
          {
            PyObject *tmp = $result;
            $result = PyList_GetItem(tmp, 0);
            Py_INCREF($result);
            Py_DECREF(tmp);
          }
          break;
      }
    }
}
```

See https://github.com/apache/subversion/blob/1.14.5/subversion/bindings/swig/include/svn_types.swg#L426-L464

**Comment by wsfulton** [comment]
@hvbtup, some guess work is required with what you've provided, as how `int *` is being handled is not clear to me. You're probably setting something that uses argout typemaps for `int *` which is the area affected by #2905. You probably need to use a custom typemap to revert to the previous behaviour as documented in the CHANGES.current/CHANGES file:

https://github.com/swig/swig/blob/8fd4df01bd4111e5f7322fe4f9d4817ffd6f0c6d/CHANGES#L377-L397

so call `SWIG_Python_AppendOutput` and set 1 for the is_void parameter.

**Comment by hvbtup** [comment]
@wsfulton I don't understand your answer. In the description I showed the relevant code. As you can see, I did not use `SWIG_Python_AppendOutput` in there, and I simply don't know how I could change my code to call that function now. 
Please keep in mind that I wrote the code ~ 20 years ago and I'm not an expert in writing SWIG typemaps.

**Comment by wsfulton** [comment]
@hvbtup the code you provided is not enough to understand why your code generates `SWIG_Python_AppendOutput`. I suggest you provide a standalone cutdown example to replicate the problem in order for you to understand and get more explicit help.

**Comment by wsfulton** [comment]
> Otherwise, I suggest just adding this as a breaking change to the release notes.

It is documented as a potentially incompatible change, see the CHANGES file https://github.com/swig/swig/blob/v4.3.0/CHANGES.current.

**Comment by hvbtup** [comment]
> It is documented as a potentially incompatible change, ...

I could be more clear perhaps. E.g. a link to this issue, including its title. 
In my case I am not using any of the `SWIG_*_AppendOutput` output functions directly in the `*.i` file, so the remarks regarding these functions did not catch my eye.

**Comment by StefanBruens** [comment]
I think this is a regression which should be fixed, or, more specifically, a convenient way to handle it should be provided.

Take this fairly common pattern:

1. Multiple functions, taking 0 or more OUTPUT parameters
2. Errors are signaled using the function return value, which may be pointers (see the SVN case above), or plain integers
3. Expected (python) signature is  `[output1, output2, ...] = func(input)`, errors are reported as python exceptions. 

Prior to this change, this was possible with an `%expception` (handling any error codes, setting `PyErr_...`) and `typemap(out)` (suppressing the return value, pushing `Py_None` into `$result`). The old AppendOutput would discard this `Py_None` in the first position for functions with 1 or more output parameters, but keep the required non-null `Py_None` object for functions without output parameters.

**Comment by StefanBruens** [comment]
Here is a testcase, including the required `%typemap(ret)` to work correctly for functions without output parameters:
```
%module test

%include "typemaps.i"

%define %MyException(funcname)
%exception funcname {
    $action
    if (result < 0) {
        PyErr_Format(PyExc_RuntimeError, "Error returned from (funcname): %ld", (long)result);
        SWIG_fail;
    }
}
%enddef

%MyException(myFunc1)
%MyException(myFunc2)

typedef int MyErr;

%typemap(out) MyErr {
    // Suppress MyErr output, error codes already handled by %expection
}

%typemap(ret) MyErr {
    // In case the wrapped function has not OUTPUT parameters result
    // is a nullptr, push the required None object
    if (!$result) {
        $result = Py_None;
        Py_INCREF($result);
    }
}

%apply char* OUTPUT { char* out1, char* out2 };
MyErr myFunc1(int i, char* out1, char* out2);
MyErr myFunc2(int i);
```

**Comment by StefanBruens** [comment]
For Swig 4.2 and older, the following `typemap` was sufficient (everything else unchanged):
```
%typemap(out) MyErr {
    // push the required None object
    $result = Py_None;
    Py_INCREF($result);
}
```

**Comment by wsfulton** [comment]
I have another alternative to restore the old behaviour without the typemap change that @StefanBruens suggests (which I think is a quite a good approach for this particular case).

Add the following to your interface file:
```swig
%{
#if SWIG_VERSION >= 0x040300
// Force third parameter to 1 to replicate swig < 4.3 behaviour
#define SWIG_Python_AppendOutput(result, obj, isvoid)  SWIG_Python_AppendOutput(result, obj, 1)
#endif
%}
```
It will be generated into the _wrap.cxx code **after** the definition of `SWIG_Python_AppendOutput` and **before** the generated wrapper functions call it. So even if the generated code passes 0 for the 3rd `is_void` parameter to `SWIG_Python_AppendOutput`, it will force it to 1. So any use of `$isvoid` in the typemaps will effectively be ignored and forced to 1. This is useful for the SWIG provided typemaps which use `$isvoid` as users modifying all these is not a great idea. However,  anyone who has their own argout typemaps that call the append functions ought to modify their own typemaps.

**Comment by wsfulton** [comment]
SWIG 4.6.0 (not released at time of writing) adds a supported way to do this: `numoutputs=0` on the `out` typemap says the function
return value is not returned to Python at all, so there is no `None` for the `OUTPUT` values to be
appended to - no dummy `Py_None` and no `SWIG_Python_AppendOutput` redefinition needed.

```swig
%typemap(out, numoutputs=0) MyErr {
    // Suppress MyErr output, error codes already handled by %expection
}
%typemap(ret) MyErr %{
  if ($1 != 0) {
    // turn the error code into an exception
    SWIG_fail;
  }
%}
```

`myFunc1(10)` then returns `[11, 12]` and `myFunc2(0)` returns `None`, as they did before 4.3. An
`argout` typemap that only raises and returns nothing of its own, like the `PTError*` ones in the
original report, should declare `numoutputs=0` too.
